add tech preview play argument spec validation

s-hertel opened the pull request:
##### SUMMARY

Play argument spec validation runs after the play-level fact gathering.

The validate_argspec keyword supports templating for parity with the gather_* keywords. It can be a bool or argument spec string.

The task is tagged always, and is skippable since it inherits play tags.

Undocumented variables are ignored by default. module_defaults can define the optional argument "provided_arguments" to ensure there's documentation for arbitrary options. This provides flexibility and is the same mechanism supporting play-level fact gathering configuration.

When validate_argspec is set to True:
  - A play name is required.
  
When validate_argspec is True or an argument spec:
  - A file in the playbook directory matching the pattern {playbook}.meta{ext?} is required.
  - The playbook meta file must contain a dictionary containing an argument spec in the format:

```yaml
argument_specs:
  Play Name:
    options: {}
 ```

##### ISSUE TYPE

- Feature Pull Request

s-hertel committed:
Add play keyword validate_argspec which injects an implicit validate_argument_spec task.

The validate_argspec supports templating for parity with the gather_* keywords.

Play argument spec validation runs after the play-level fact gathering.

Undocumented variables ignored by default.

module_defaults can define the optional argument "provided_arguments"
to ensure there's documentation for arbitrary options.
This provides flexibility and is the same mechanism supporting play-level
fact gathering configuration.

When validate_argspec is set to True:
  - A play name is required.
  - A file in the playbook directory matching the pattern {playbook}.meta{ext?} is required.
  - The playbook meta file must contain a dictionary containing an argument spec for the
    play such as {"argument_specs": {"Play Name": {"options": {}}}.

s-hertel committed:
improve tests

s-hertel committed:
remove superfluous line and update code comment

Shrews commented on changelogs/fragments/play-argument-spec-validation.yml:
```diff
@@ -0,0 +1,7 @@
+  enabled by setting the play keyword ``validate_argspec`` to ``True``.
+  This will load a file matching the pattern <playbook_name>.meta.yml,
+  which should contain a top dictionary "argument_specs", containing
+  the play name, containing an "options" dictionary.
```
IIRC, for role argspec validation, the existence of an argspec file means we automatically do the validation. I think `include_role` and `import_role` have options to disable. Perhaps we follow the same pattern for consistency sake? If an argspec exists and a play name matches, always do the validation. The new option could then be used to explicitly disable.

s-hertel replied:
Good point, I wouldn't be opposed to that. The design document specified that the existence of a play argument spec should not enable validation on its own, which is why I did it this way. @sivel and @nitzmahone What are your thoughts on this suggestion?

sivel replied:
The goal was to make it explicitly opt in, to also catch the case of a play being renamed, which is potentially more likely than a role file being renamed. If it were explicitly enabled, you would know when the argspec didn't match because of getting an error.

On the flip side, renaming the play would just result in no argspec validation happening.

I think we want to stick with an explicit enable.

s-hertel replied:
@sivel That makes sense. The role task entrypoints would need to be renamed, just renaming the role wouldn't break the role argument spec validation. Whereas implicit play argument spec validation would break by renaming the playbook or play, or even by providing different extra vars if play names are dynamic. I think I like the simplicity of requiring opt-in.

What about allowing explicit opt-in to enforce validation, but implicitly validate arguments if the keyword is omitted and a play argument spec exists? Maybe this would cause more confusion/risk than the convenience it provides. :thinking:

sivel replied:
> What about allowing explicit opt-in to enforce validation, but implicitly validate arguments if the keyword is omitted and a play argument spec exists? Maybe this would cause more confusion/risk than the convenience it provides.

We can always expand this later in future versions if it seems like users are interested.

mkrizek commented on lib/ansible/executor/play_iterator.py:
```diff
@@ -271,35 +292,36 @@ def _get_next_task_from_state(self, state, host):
-                # the specified host.
-                if not state.pending_setup:
-                    state.pending_setup = True
+                # First, we check to see if we completed both setup tasks injected
```
After staring at this for a while the `PlayIterator` changes look good to me.

s-hertel replied:
@mkrizek Thank you for the review!

s-hertel committed:
Simplify tag handling - inherit play tags

Add tests for the interaction between validate_argspec and tags

Simplify test variables

sivel commented on lib/ansible/playbook/play.py:
```diff
@@ -17,6 +17,10 @@
 
+import functools as _functools
+import pathlib as _pathlib
+import yaml as _yaml
```
Not sure if it is worth it, but we have `ansible.module_utils.common.yaml::yaml_dump` which will make use of the libyaml extensions.

sivel commented on lib/ansible/playbook/play.py:
```diff
@@ -64,6 +68,8 @@ class Play(Base, Taggable, CollectionSearch):
     vars_files = NonInheritableFieldAttribute(isa='list', default=list, priority=99)
     vars_prompt = NonInheritableFieldAttribute(isa='list', default=list, always_post_validate=False)
 
+    validate_argspec = NonInheritableFieldAttribute(isa='bool', default=False, always_post_validate=True)
```
There was a late addition/comment to the architecture doc, that proposed making this bool or the name of an argspec, that wasn't tied directly to the play name. Do you have any thoughts or feelings on that?

s-hertel replied:
I think it's a good suggestion, I'll adjust that.  It is more convenient to only need to define one keyword instead of two. If a play name and a validate_argspec name are both provided, which should win? Or should that be an error?

I'm also wondering if there are any better names for this option.

* `playspec_validate` would be more similar to the include/import_role toggle `rolespec_validate`
* `validate_argument_spec` would be easily associated with the action

sivel replied:
> If a play name and a validate_argspec name are both provided, which should win? Or should that be an error?

It should just use the provided name for the `validate_argspec`. I see no reason this should be an error, it would also allow sharing easily.


> I'm also wondering if there are any better names for this option.
> 
> * `playspec_validate` would be more similar to the include/import_role toggle `rolespec_validate`
> * `validate_argument_spec` would be easily associated with the action

If it's changed, I'd go for more verbose rather than less, and call it `validate_argument_spec`, but I'm also ok with the current name.

s-hertel replied:
I've added this. If anyone feels strongly about the name I'll change that too, but I've left it as-is for now.

s-hertel replied:
@sivel @nitzmahone @bcoca I've left this thread unresolved to make sure we reach a consensus.

If we do not want to add a singular true|false|string option, I think we should just add a string option so users can enable the feature with 1 keyword always, instead of requiring a play name under some circumstances.

sivel replied:
Ok, there has been a bit of discussion about this in various places.

@nitzmahone and I had a conversation, and we're going to stick with the implmentation and what was instructed in the architecture doc.  Specifically the single keyword with `true/false/string`

s-hertel committed:
Apply suggested changes

The validate_argspec name is now configurable as a argument spec name or boolean

Use module_utils for dumping yaml

Refactor tests for legibility

s-hertel committed:
update changelog

s-hertel committed:
validation is skipped when validate_argspec is False below

s-hertel committed:
Fix handling missing play names, add tests

Include full playbook meta candidates in the error

s-hertel committed:
Test setting validate_argspec to no versus 'no'

sivel commented on lib/ansible/playbook/play.py:
```diff
@@ -437,3 +445,92 @@ def copy(self):
+        else:
+            playbook_without_ext = self._origin.path
+
+        return [_pathlib.Path(playbook_without_ext + ".meta" + ext) for ext in extensions + ['']]
```
Totally not critical, but we could restrict this to match the extension of the origin, and not look for others.

s-hertel replied:
The playbook extension is optional (which surprised me) and I thought we should still prefer an extension in that case. I considered hardcoding `.yml` because we do that for the `runtime.yml` (and maybe in another spot or two), but the role argument spec filename handling works with `YAML_FILENAME_EXTENSIONS` and it didn't add much complexity imo.

sivel commented on lib/ansible/playbook/play.py:
```diff
@@ -437,3 +445,92 @@ def copy(self):
+            metadata = metadata[argspec_name]
+            options = metadata['options']
+        except (TypeError, KeyError):
+            options = None
```
Not required for now, could be for later, but since we load the file here, should we cache it as `self._argument_spec` instead of needing to load it again  via `PlayIterator`?  The property below could then return the cache.  Just a thought.

s-hertel replied:
The DataLoader.load_from_file uses a cache, which is why I didn't add one here, since the property below can use the cache. We have so many layers of caches, so I tried to add them strategically. I cached `_metadata_candidate_paths` and `_metadata_path` so the same file is used.

sivel replied:
Doh, yep, you are right, I was forgetting a lower implementation detail.

s-hertel commented:
Thanks for the reviews and discussion!
